### PR TITLE
Config center feature implement for setting the zookeeper as config center to refresh consumerConfig & providerConfig when dubbogo starting.

### PR DESCRIPTION
**Pull request by hxmhlt**
**What this PR does**:

New feature implement for config center,  config refresh from config center when dubbogo start.
Now just have a zookeeper implement for config center.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
You can do a example in examples/dubbo/with-configcenter-go-server/

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by wongoo** [comment]
how about make a conference to introduce it to reviewers when there is a big exchange like this?

**Comment by hxmhlt** [comment]
> how about make a conference to introduce it to reviewers when there is a big change like this?

ok, How can we introduce it to reviewers?  Like a meeting ? Or an introduction article ?

**Comment by hxmhlt** [comment]
1.Code in config folder have refactored for supporting config_center to refresh the consumerConfig or providerConfig when dubbogo starting. The core code is in  base_config.go.
2.Interface in config_center defined and I implement zookeeper as config_center ,also make it compatible with java 2.7.2.
3.TODO: Dynamic config by using Listeners's in dynamicConfigCenter. Maybe next pr.

**Comment by hxmhlt** [comment]
> how about make a conference to introduce it to reviewers when there is a big change like this?

I write the main work in this pr above . May can help you to review.

File: common/config/environment.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+type Environment struct {
+	configCenterFirst bool
+	externalConfigs   sync.Map
+	externalConfigMap sync.Map
+}
+
+var instance *Environment
+
+var once sync.Once
+
```
Comment (AlexStocks) [inline common/config/environment.go]: 这两个变量放在一个var () 里面吧

File: common/config/environment.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	conf.store = p
+}
+
+func (conf *InmemoryConfiguration) GetProperty(key string) (bool, string) {
+	v, ok := conf.store.Load(key)
+	if ok {
+		return true, v.(string)
+	} else {
+		return false, ""
+	}
```
Comment (AlexStocks) [inline common/config/environment.go]: 直接写成 
v, ok := conf.store.Load(key)
if ok {
	return true, v.(string)
} 
		
return false, ""

File: common/constant/default.go [hunk 3]
```diff
@@ -17,6 +17,7 @@
 
 package constant
 
+const DUBBO = "dubbo"
```
Comment (AlexStocks) [inline common/constant/default.go]: 上下风格统一，都放在一个 const () 里面

File: common/extension/config_center_factory.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+func SetConfigCenterFactory(name string, v func() config_center.DynamicConfigurationFactory) {
+	configCenterFactories[name] = v
+}
+
+func GetConfigCenterFactory(name string) config_center.DynamicConfigurationFactory {
+	if configCenterFactories[name] == nil {
+		panic("config center for " + name + " is not existing, make sure you have import the package.")
+	}
+	return configCenterFactories[name]()
+
```
Comment (AlexStocks) [inline common/extension/config_center_factory.go]: 最后一行为空不好吧

File: config/base_config.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	if err != nil {
+		return perrors.WithStack(err)
+	}
+	config.GetEnvInstance().UpdateExternalConfigMap(mapContent)
+	return nil
+}
+
+func getKeyPrefix(val reflect.Value, id reflect.Value) string {
+	var prefix string
+	var idStr string
```
Comment (AlexStocks) [inline config/base_config.go]: var ()

File: config/config_loader.go [hunk 6]
```diff
@@ -66,172 +56,31 @@ func init() {
...
-var (
-	refConfigs map[string]*ReferenceConfig // record reference config loaded
-	srvConfigs map[string]*ServiceConfig   // record service config loaded
-)
-
 // Dubbo Init
-func Load() (int, int) {
+func Load() {
 	var refMap map[string]*ReferenceConfig
 	var srvMap map[string]*ServiceConfig
```
Comment (AlexStocks) [inline config/config_loader.go]: var () block

File: config_center/configuration_parser_test.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package config_center
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
```
Comment (AlexStocks) [inline config_center/configuration_parser_test.go]: 两个 import () block

File: config_center/dynamic_configuration_factory.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package config_center
+
+import "github.com/apache/dubbo-go/common"
```
Comment (AlexStocks) [inline config_center/dynamic_configuration_factory.go]: import ()

File: config_center/mock_dynamic_config.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import (
+	"sync"
+)
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+type MockDynamicConfigurationFactory struct {
+}
```
Comment (AlexStocks) [inline config_center/mock_dynamic_config.go]: 放在一行内

File: config_center/mock_dynamic_config.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+var once sync.Once
+var dynamicConfiguration *mockDynamicConfiguration
```
Comment (AlexStocks) [inline config_center/mock_dynamic_config.go]: var () block

File: protocol/jsonrpc/http.go [hunk 11]
```diff
@@ -110,7 +110,7 @@ func (c *HTTPClient) NewRequest(service common.URL, method string, args interfac
 func (c *HTTPClient) Call(ctx context.Context, service common.URL, req *Request, rsp interface{}) error {
 	// header
 	httpHeader := http.Header{}
-	httpHeader.Set("Content-Type", "application/json")
+	httpHeader.Set("Content-Protocol", "application/json")
```
Comment (AlexStocks) [inline protocol/jsonrpc/http.go]: content-type -> content-protocol 理由何在？

File: remoting/listener.go [hunk 12]
```diff
@@ -46,6 +46,7 @@ type EventType int
 const (
 	Add = iota
 	Del
+	Mod
```
Comment (AlexStocks) [inline remoting/listener.go]: Mod --> Update，因为go里面没有enum，最好再给你这三个名字前面加个前缀作为enum名称，以作区分

File: common/config/environment.go (hunk 2 above)
Comment (hxmhlt) [reply common/config/environment.go]: 这样写不是语义更明确一些吗?

File: protocol/jsonrpc/http.go (hunk 11 above)
Comment (hxmhlt) [reply protocol/jsonrpc/http.go]: @fangyincheng 这是你改的吧？

File: common/config/environment.go (hunk 2 above)
Comment (AlexStocks) [reply common/config/environment.go]: 我的更明确，你理解下

File: remoting/listener.go (hunk 12 above)
Comment (hxmhlt) [reply remoting/listener.go]: good suggestion!

File: protocol/jsonrpc/http.go (hunk 11 above)
Comment (fangyincheng) [reply protocol/jsonrpc/http.go]: @hxmhlt 我看了我的分支，不是我改的，应该你怎么操作误改了吧

File: config_center/zookeeper/impl.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zookeeper
+
+import (
+	"github.com/samuel/go-zookeeper/zk"
```
Comment (fangyincheng) [inline config_center/zookeeper/impl.go]: pkg fmt

File: protocol/jsonrpc/http.go (hunk 11 above)
Comment (hxmhlt) [reply protocol/jsonrpc/http.go]: 确实是我改的，不知道为啥了。我改回来了。

File: config/protocol_config.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package config
+
+import (
+	"github.com/apache/dubbo-go/common/constant"
+	"strings"
```
Comment (AlexStocks) [inline config/protocol_config.go]: two import blocks

File: config/method_config.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package config
+
+import "github.com/apache/dubbo-go/common/constant"
```
Comment (AlexStocks) [inline config/method_config.go]: import ()

File: config/provider_config.go [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+func GetProviderConfig() ProviderConfig {
+	if providerConfig == nil {
+		logger.Warnf("providerConfig is nil!")
+		return ProviderConfig{}
+	}
+	return *providerConfig
+}
+
+func providerInit(confProFile string) error {
+	if confProFile == "" {
```
Comment (AlexStocks) [inline config/provider_config.go]: if len(confProFile) == 0

File: config_center/configuration_parser.go [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"github.com/apache/dubbo-go/common/logger"
+)
+
+type ConfigurationParser interface {
+	Parse(string) (map[string]string, error)
+}
+
+//for support properties file in config center
+type DefaultConfigurationParser struct {
```
Comment (AlexStocks) [inline config_center/configuration_parser.go]: Just write a empty struct in one line

File: config_center/zookeeper/impl.go [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+
+	opions := &config_center.Options{}
+	for _, opt := range opts {
+		opt(opions)
+	}
+	/**
+	 * when group is not null, we are getting startup configs from Config Center, for example:
+	 * group=dubbo, key=dubbo.properties
+	 */
+	if opions.Group != "" {
```
Comment (AlexStocks) [inline config_center/zookeeper/impl.go]: if len(options.Group) == 0

File: config_center/zookeeper/impl.go [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+	c.cacheListener.AddListener(key, listener)
+}
+
+func (c *zookeeperDynamicConfiguration) RemoveListener(key string, listener remoting.ConfigurationListener, opions ...config_center.Option) {
+	c.cacheListener.RemoveListener(key, listener)
+}
+
+func (c *zookeeperDynamicConfiguration) GetConfig(key string, opts ...config_center.Option) (string, error) {
+
+	opions := &config_center.Options{}
```
Comment (AlexStocks) [inline config_center/zookeeper/impl.go]: opions = options

File: config_center/zookeeper/impl_test.go [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+	"context"
+	"fmt"
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/config_center"
+	"github.com/apache/dubbo-go/remoting"
+	"github.com/samuel/go-zookeeper/zk"
+	"github.com/stretchr/testify/assert"
+	"sync"
+	"testing"
+)
```
Comment (AlexStocks) [inline config_center/zookeeper/impl_test.go]: three import block.

File: examples/dubbo/go-client/app/client.go [hunk 21]
```diff
@@ -57,10 +57,7 @@ func main() {
 	hessian.RegisterJavaEnum(Gender(WOMAN))
 	hessian.RegisterPOJO(&User{})
 
-	conLen, _ := config.Load()
-	if conLen == 0 {
-		panic("conMap is nil")
-	}
+	config.Load()
```
Comment (AlexStocks) [inline examples/dubbo/go-client/app/client.go]: why delete check condition?

File: examples/dubbo/go-server/profiles/release/server.yml [hunk 22]
```diff
@@ -1,52 +1,49 @@
...
   version : "0.0.1"
   owner : "ZX"
-  environment : "dev"
+  environment : "release"
 
 registries :
-  - id: "hangzhouzk"
-    type: "zookeeper"
+  "hangzhouzk":
+    protocol: "zookeeper"
```
Comment (AlexStocks) [inline examples/dubbo/go-server/profiles/release/server.yml]: type -> protocol? Is zookeeper  a protocol?

File: remoting/zookeeper/facade.go [hunk 23]
```diff
@@ -30,7 +30,7 @@ import (
 	"github.com/apache/dubbo-go/common/logger"
 )
 
-type ZkClientContainer interface {
+type zkClientFacade interface {
```
Comment (AlexStocks) [inline remoting/zookeeper/facade.go]: which change its name?

File: examples/dubbo/go-client/app/client.go (hunk 21 above)
Comment (hxmhlt) [reply examples/dubbo/go-client/app/client.go]: config.Load方法已经改为无返回值。之前是返回加载成功的配置数，与yincheng讨论后觉得没有必要，且不优雅。

File: examples/dubbo/go-server/profiles/release/server.yml (hunk 22 above)
Comment (hxmhlt) [reply examples/dubbo/go-server/profiles/release/server.yml]: 是的，这里与java的叫法保持一致了。

Comment (hxmhlt) [reply examples/dubbo/go-server/profiles/release/server.yml]: 是的，这里与java的叫法保持一致了。

File: remoting/zookeeper/facade.go (hunk 23 above)
Comment (hxmhlt) [reply remoting/zookeeper/facade.go]: container叫法感觉别扭，感觉这种实现方式，类似于设计模式之门面模式。

File: common/config/environment.go [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+func GetEnvInstance() *Environment {
+	once.Do(func() {
+		instance = &Environment{configCenterFirst: true}
+	})
+	return instance
+}
+
+//func (env *Environment) SetConfigCenterFirst() {
```
Comment (gaoxinge) [inline common/config/environment.go]: Why not remove comment?

File: common/config/environment.go [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+//	env.configCenterFirst = true
+//}
+
+//func (env *Environment) ConfigCenterFirst() bool {
```
Comment (gaoxinge) [inline common/config/environment.go]: As above.

File: config_center/zookeeper/impl_test.go [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+ */
+package zookeeper
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+)
+import (
```
Comment (gaoxinge) [inline config_center/zookeeper/impl_test.go]: Better add a space line between two imports.

File: config_center/zookeeper/impl_test.go [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/samuel/go-zookeeper/zk"
+	"github.com/stretchr/testify/assert"
+)
+import (
```
Comment (gaoxinge) [inline config_center/zookeeper/impl_test.go]: As above.

File: config/provider_config.go [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package config
+
+import (
+	"context"
+	"github.com/apache/dubbo-go/common/constant"
```
Comment (gaoxinge) [inline config/provider_config.go]: Move to third import clause.

File: config/provider_config.go [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+	"io/ioutil"
+	"path"
+)
+
+import (
+	perrors "github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+import (
```
Comment (gaoxinge) [inline config/provider_config.go]: Better add a space line between two imports.

File: common/url.go [hunk 30]
```diff
@@ -197,10 +204,10 @@ func NewURL(ctx context.Context, urlString string, opts ...option) (URL, error)
 	//		s.Timeout = time.Duration(timeout * 1e6) // timeout unit is millisecond
 	//	}
 	//}
+	//fmt.Println(s.String())
```
Comment (gaoxinge) [inline common/url.go]: Remove comment.

**Comment by gaoxinge** [comment]
I have added some review comments on codes. I also want to ask when this pr will be merged into master?

**Comment by AlexStocks** [comment]
> I have added some review comments on codes. I also want to ask when this pr will be merged into master?

I do not know. So big pr.

File: common/config/environment.go (hunk 25 above)
Comment (hxmhlt) [reply common/config/environment.go]: 这里我估计后面添加新的配置级别，就会要加上，所以没有直接去掉，先注释掉了。先不删吧

**Pull request merged**